node/src/node_configurator/mod.rs

```diff
@@ -11,34 +11,218 @@ use crate::database::db_initializer::{DbInitializer, DbInitializerReal};
+                                false => match data_directory_def {
+                                    true => PathBuf::from(data_directory).join(PathBuf::from(config_str)),
+                                    false => panic!("You need to define data-directory to define config file with naked directory 'dirname/config.toml'.")
+                                }
+                            }
+                        }
+                    }
+                }
+                false => PathBuf::from(config_str),
+            };
+            (path, true)
+        }
+        None => {
+            let path = PathBuf::from(data_directory).join(PathBuf::from("config.toml"));
+            match path.is_file() {
+                true => (path, true),
+                false => (path, false),
+            }
+        }
+    }
+}
+
+fn config_file_data_dir_real_user_chain_from_vcl(
+    dirs_wrapper: &dyn DirsWrapper,
+    configs: Vec<Box<dyn VclArg>>,
+) -> UserSpecifiedData {
+    let mut user_specified_data = UserSpecifiedData {
+        chain: Default::default(),
+        chain_spec: false,
+        real_user: Default::default(),
+        real_user_spec: false,
+        data_directory: Default::default(),
+        data_directory_spec: false,
+        config_file: Default::default(),
+        config_file_spec: false,
+    };
+    let configs = configs.as_slice();
+    (user_specified_data.chain, user_specified_data.chain_spec) = get_chain_from_vcl(configs);
+    (
+        user_specified_data.real_user,
+        user_specified_data.real_user_spec,
+    ) = get_real_user_from_vcl(configs, dirs_wrapper);
+    (
+        user_specified_data.data_directory,
+        user_specified_data.data_directory_spec,
+    ) = get_data_directory_from_vcl(
+        configs,
+        dirs_wrapper,
+        &user_specified_data.real_user,
+        &user_specified_data.chain,
+    );
+    (
+        user_specified_data.config_file,
+        user_specified_data.config_file_spec,
+    ) = get_config_file_from_vcl(
+        configs,
+        &user_specified_data.data_directory,
+        user_specified_data.data_directory_spec,
+        dirs_wrapper,
+    );
+    user_specified_data
+}
+
+fn argument_from_vcl<'a>(configs: &'a [Box<dyn VclArg>], needle: &'a str) -> Option<&'a str> {
+    match configs
+        .iter()
+        .find(|vcl_arg_box| vcl_arg_box.deref().name() == needle)
+    {
+        Some(vcl_arg_box) => vcl_arg_box.deref().value_opt(),
+        None => None,
+    }
+}
+
+pub fn determine_user_specific_data(
     dirs_wrapper: &dyn DirsWrapper,
     app: &App,
     args: &[String],
-) -> Result<(PathBuf, bool, PathBuf, RealUser), ConfiguratorError> {
-    let orientation_schema = App::new("MASQNode")
-        .arg(chain_arg())
-        .arg(real_user_arg())
-        .arg(data_directory_arg(DATA_DIRECTORY_HELP))
-        .arg(config_file_arg());
+) -> Result<UserSpecifiedData, ConfiguratorError> {
     let orientation_args: Vec<Box<dyn VclArg>> = merge(
```

Comment:
I think this should be replaced with a full-fledged MultiConfig, and that the filtering I put in years ago is superfluous and should be removed. I think the filtering reveals some intention, and might be important in that respect, but MultiConfig isn't designed to do such a thing (and shouldn't be, in my opinion), so although the intention is revealed, the code is necessarily obtuse and confusing. Intention could probably be revealed just as effectively with a comment declaring that only four fields are of interest.